### PR TITLE
deprecate vars: handle init-less vars, bail out to var in scoped usage, basic destructuring support

### PR DESCRIPTION
**Pull request by iamdustan**
To start cleaning up the deprecation of vars. Also, we ran this over some local projects that had already gone `let` so I removed the “must-be-vars” filter to allow for lets to become const as appropriate as well.

### CODE REVIEW

**Comment by iamdustan** [comment]
Not handling any of @spicyj’s variable hoisting scenarios yet, but this does handle init-less declarations as well as multiple-declarators.

**Comment by iamdustan** [comment]
Okay okay okay, I added support for bailing out to `var`s in case of loop-ed new scope access in 94dcb69.

File: transforms/no-vars.js [hunk 1]
```diff
@@ -49,24 +86,40 @@ module.exports = function(file, api) {
...
           'ForStatement' === p.parent.value.type ||
           'ForInStatement' === p.parent.value.type ||
-          'ForOfStatement' === p.parent.value.type ||
-          isMutated(p)
+          'ForOfStatement' === p.parent.value.type
         ) {
-          p.value.kind = 'let';
+          if (!isAccessedInClosure(p.parent)) {
+            p.value.kind = 'let';
+          }
```
Comment (iamdustan) [inline transforms/no-vars.js]: do we want to log and warn the user about the potential race condition in this case?

**Comment by iamdustan** [comment]
Also basic destructuring support. I’m really done for now. Full destructuring support is still needed as well as var hoisting stuff, but that can be another PR :)

**Comment by iamdustan** [comment]
Well snap! I just ran it over Relay and `npm test` *almost* passed! It is hanging on one test. I’m suspecting currently that it has to do with [src/store/RelayStoreGarbageCollector.js](https://github.com/facebook/relay/blob/master/src/store/RelayStoreGarbageCollector.js#L106)

At first glance it looks likely that the transform could be causing some subtle changes in behavior here, but not sure yet.

**Comment by iamdustan** [comment]
Oh yeah, the Relay transformation shortstat.  200 files changed, 5633 insertions(+), 5643 deletions(-). No biggie.

**Comment by iamdustan** [comment]
most of the differences in insertions/deletions are purely formatting differences as well.

**Comment by cpojer** [comment]
wtf. I had notifications for this repo off. Awesome, I'll take a look tomorrow.

File: transforms/no-vars.js [hunk 2]
```diff
@@ -20,26 +20,87 @@ module.exports = function(file, api) {
...
+      .find(j.Identifier)
+      .filter(n => {
+        const declarations = node.value.init ? node.value.init.declarations
+                           : node.value.left ? node.value.left.declarations
+                           : [];
+        if (declarations.some(d => d.id.name === n.value.name)) {
+          let parent = n.parent;
+          while (parent.value !== node.value.body) {
+            parent = parent.parent;
+            if (/Function/.test(parent.value.type)) {
```
Comment (cpojer) [inline transforms/no-vars.js]: this would be nice if it was actually checking against string values directly.

File: transforms/no-vars.js (hunk 1 above)
Comment (cpojer) [inline transforms/no-vars.js]: yeah I think that would be useful.

File: transforms/no-vars.js [hunk 3]
```diff
@@ -49,24 +110,45 @@ module.exports = function(file, api) {
...
         ) {
-          p.value.kind = 'let';
+          if (!isAccessedInClosure(p.parent)) {
+            p.value.kind = 'let';
+          }
           return true;
         } else {
-          p.value.kind = 'const';
-          return true;
+          const lets = [], consts = [];
```
Comment (cpojer) [inline transforms/no-vars.js]: please two-line this :)

**Comment by cpojer** [comment]
I think this is a good continuation of the previous PR. Can you clean up those two minor things so I can merge it?

File: transforms/no-vars.js (hunk 2 above)
Comment (iamdustan) [inline transforms/no-vars.js]: I wasn’t sure what all the `/Function/` style ASTs I needed to check against off-hand. E.g. FunctionExpression, FunctionDeclaration, GeneratorFunction, AsyncFunctionExpression?

**Comment by cpojer** [comment]
One concern I heard from @sebmck was that this might spit out some code that is invalid with regards to the TMZ (although, I will assume, that running this code through babel would point that out). So the other solution to this codemod could be to use babel and babel's static analysis tools to modify the AST and then use recast to print it back.

File: transforms/no-vars.js (hunk 2 above)
Comment (cpojer) [inline transforms/no-vars.js]: you can find the types here: https://github.com/benjamn/ast-types/blob/master/def/

**Comment by iamdustan** [comment]
> One concern I heard from @sebmck was that this might spit out some code that is invalid with regards to the TMZ

I’ve just been pretending that doesn’t exist. :) I think that the var hoisting issues (that have not been handled at all yet) and TMZ are very much in the same vein. What would the recast/babel integration look like?

Also, the things have been cleaned up. Attached is a screenshot of the output from one of the test cases that bails out to the `var`.

screen shot 2015-10-26 at 6 48 36 am

**Comment by cpojer** [comment]
yay!

**Comment by sebmck** [comment]
Just a nit pick but TMZ is a celebrity news website/TV show while TDZ (temporal dead zone) is the JavaScript construct :stuck_out_tongue:

**Comment by cpojer** [comment]
lol.

**Comment by iamdustan** [comment]
I would say that I fat-fingered that but clearly M and D are nowhere near each other. :innocent:

**Pull request merged**